src/librustc_trans/cabi_mips64.rs

```diff
@@ -8,50 +8,165 @@
     let dl = &cx.tcx.data_layout;
     let size = arg.layout.size;
-    let align = arg.layout.align.max(dl.i32_align).min(dl.i64_align);
 
     if arg.layout.is_aggregate() {
-        arg.cast_to(Uniform {
-            unit: Reg::i64(),
-            total: size
-        });
-        if !offset.is_abi_aligned(align) {
-            arg.pad_with(Reg::i64());
+        let mut chunks: Vec<Reg> = Vec::new();
+        let trailing_int_size = match arg.layout.fields {
+            layout::FieldPlacement::Array { .. } => {
+                // Arrays are passed indirectly
+                arg.make_indirect();
+                return;
+            }
+            layout::FieldPlacement::Union(_) => {
+                // Unions and are always treated as a series of 64-bit integer chunks
+                size
+            },
+            layout::FieldPlacement::Arbitrary { .. } => {
+                // Structures are split up into a series of 64-bit integer chunks, but any aligned
+                // doubles not part of another aggregate are passed as floats.
+                let mut last_offset = Size::from_bytes(0);
+
+                for i in 0..arg.layout.fields.count() {
+                    let field = arg.layout.field(cx, i);
+                    let offset = arg.layout.fields.offset(i);
+
+                    // We only care about aligned doubles
+                    if let layout::Abi::Scalar(ref scalar) = field.abi {
+                        if let layout::F64 = scalar.value {
+                            if offset.is_abi_aligned(dl.f64_align) {
+                                // Insert enough integers to cover [last_offset, offset)
+                                assert!(last_offset.is_abi_aligned(dl.f64_align));
+                                for _ in 0..((offset - last_offset).bits() / 64) {
+                                    chunks.push(Reg::i64());
+                                }
+
+                                chunks.push(Reg::f64());
+                                last_offset = offset + Reg::f64().size;
+                            }
+                        }
+                    }
+                }
+
+                size - last_offset
+            }
+        };
+
+        // If there are no doubles in this structure, simplify to Uniform
+        if chunks.is_empty() {
+            arg.cast_to(Uniform {
+                unit: Reg::i64(),
+                total: size
+            });
+        } else {
+            // Add any trailing integers. These are always I64 chunks except for the last
+            // chunk which might be a smaller integer.
+            for _ in 0..(trailing_int_size.bits() / 64) {
+                chunks.push(Reg::i64());
+            }
+
+            let final_int_bits = trailing_int_size.bits() % 64;
+            if final_int_bits > 0 {
+                chunks.push(bits_to_int_reg(final_int_bits));
+            }
+
+            arg.cast_to(CastTarget::Chunked(chunks));
```

Comment:
Ah, you don't need this! This is literally what `Uniform { unit: Reg::i64(), total: trailing_int_size }` does!

Reply:
I think it's only equivalent to `Uniform` if `chunks` is empty when entering the else block.

Reply:
Ah sorry I didn't see it came from more code above 😞.